src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
             return -1;
         }
 
+        public static void Replace<T>(Span<T> span, T oldValue, T newValue)
+        {
+            for (int i = 0; i < span.Length; ++i)
+            {
+                ref T val = ref span[i];
+
+                if (EqualityComparer<T>.Default.Equals(oldValue, val))
+                {
+                    val = newValue;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ReplaceValueType<T>(ref T src, ref T dst, T oldValue, T newValue, nuint length) where T : struct
+        {
+            Debug.Assert(Vector128.IsHardwareAccelerated && Vector128<T>.IsSupported);
+
+            if (length < (uint)Vector128<T>.Count)
+            {
+                for (nuint idx = 0; idx < length; ++idx)
+                {
+                    T original = Unsafe.Add(ref src, idx);
+                    Unsafe.Add(ref dst, idx) = EqualityComparer<T>.Default.Equals(original, oldValue) ? newValue : original;
+                }
+            }
+            else
+            {
+                ReplaceValueTypeVectorized(ref src, ref dst, oldValue, newValue, length);
```

Comment:
I split the vectorized code-path into an extra method (that won't be inlined), so that the scalar-path is tight and fast. Benchmarking for string.Replace showed that this is needed to be on par.

Reply:
Does this mean the scalar loop is going to be inlined into every MemoryExtensions.Replace call site?  That doesn't seem ideal.

Reply:
Yes and yes, so I changed (back) to a non-annoted method without the split. 
Benchmark for string.Replace looks now like:
```
|         Method | Length |      Mean |    Median | Ratio |
|--------------- |------- |----------:|----------:|------:|
|        Default |      7 |  29.04 ns |  28.99 ns |  1.00 |
| ViaSpanHelpers |      7 |  28.90 ns |  28.50 ns |  0.94 |
|                |        |           |           |       |
|        Default |      8 |  30.63 ns |  28.76 ns |  1.00 |
| ViaSpanHelpers |      8 |  21.92 ns |  21.76 ns |  0.75 |
|                |        |           |           |       |
|        Default |     15 |  39.29 ns |  39.15 ns |  1.00 |
| ViaSpanHelpers |     15 |  25.06 ns |  24.82 ns |  0.64 |
|                |        |           |           |       |
|        Default |     16 |  25.84 ns |  24.10 ns |  1.00 |
| ViaSpanHelpers |     16 |  27.48 ns |  27.29 ns |  1.09 |
|                |        |           |           |       |
|        Default |     31 |  26.17 ns |  25.56 ns |  1.00 |
| ViaSpanHelpers |     31 |  29.72 ns |  29.31 ns |  1.12 |
|                |        |           |           |       |
|        Default |    100 |  47.76 ns |  46.54 ns |  1.00 |
| ViaSpanHelpers |    100 |  39.16 ns |  38.36 ns |  0.80 |
|                |        |           |           |       |
|        Default |    500 | 108.94 ns | 105.87 ns |  1.00 |
| ViaSpanHelpers |    500 | 119.06 ns | 110.75 ns |  1.10 |
|                |        |           |           |       |
|        Default |   1000 | 206.77 ns | 195.29 ns |  1.00 |
| ViaSpanHelpers |   1000 | 187.44 ns | 184.15 ns |  0.92 |
```
Numbers especially for higher length are flaky. Run multiple times, turned memory-randomizaiton on and off. Sometimes it looks like shown, sometimes e.g. length 1000 is slower.

I don't like that split into methods for scalar / vectorized neither -- it's not natural.
So I pushed the change and then lets see what perf-lab says?